components/net/http_cache.rs

```diff
@@ -619,24 +624,24 @@ impl HttpCache {
         let entry_key = CacheKey::new(request.clone());
         if let Some(cached_resources) = self.entries.get_mut(&entry_key) {
             for cached_resource in cached_resources.iter_mut() {
-                let mut stored_headers = cached_resource.metadata.headers.lock().unwrap();
                 // Received a response with 304 status code, in response to a request that matches a cached resource.
                 // 1. update the headers of the cached resource.
                 // 2. return a response, constructed from the cached resource.
-                stored_headers.extend(response.headers.iter());
-                let mut constructed_response = Response::new(cached_resource.metadata.final_url.clone());
-                constructed_response.headers = stored_headers.clone();
+                let mut constructed_response = Response::new(cached_resource.data.metadata.data.final_url.clone());
                 constructed_response.body = cached_resource.body.clone();
-                constructed_response.status = cached_resource.status.clone();
-                constructed_response.https_state = cached_resource.https_state.clone();
+                constructed_response.status = cached_resource.data.status.clone();
+                constructed_response.https_state = cached_resource.data.https_state.clone();
                 constructed_response.referrer = request.referrer.to_url().cloned();
                 constructed_response.referrer_policy = request.referrer_policy.clone();
-                constructed_response.raw_status = cached_resource.raw_status.clone();
-                constructed_response.url_list = cached_resource.url_list.clone();
+                constructed_response.raw_status = cached_resource.data.raw_status.clone();
+                constructed_response.url_list = cached_resource.data.url_list.clone();
                 // done_chan will have been set to Some by http_network_fetch,
                 // set it back to None since the response returned here replaces the 304 one from the network.
                 *done_chan = None;
-                cached_resource.expires = get_response_expiry(&constructed_response);
+                cached_resource.data.expires = get_response_expiry(&constructed_response);
+                let mut stored_headers = cached_resource.data.metadata.headers.lock().unwrap();
```

Comment:
Why were these lines rearranged, out of curiosity?

Reply:
I got this error:
```
error[E0502]: cannot borrow `cached_resource.data` as mutable because it is also borrowed as immutable
   --> components/net/http_cache.rs:644:17
    |
627 |                 let mut stored_headers = cached_resource.data.metadata.headers.lock().unwrap();
    |                                          -------------------- immutable borrow occurs here
...
644 |                 cached_resource.data.expires = get_response_expiry(&constructed_response);
    |                 ^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
645 |                 return Some(constructed_response);
646 |             }
    |             - immutable borrow ends here
```
Maybe I could have called `.clone()` instead of moving it around?

Reply:
Ah, that makes sense. Cloning would not have been the right solution, since the original value would not have been updated. This solution seems fine to me.